lightning-block-sync/src/http.rs

```diff
@@ -0,0 +1,765 @@
+					*self = Self::connect(endpoint)?;
+					self.send_request(request).await
+				},
+				_ => Err(e),
+			},
+		}
+	}
+
+	/// Sends an HTTP request message and reads the response, returning its body.
+	async fn send_request(&mut self, request: &str) -> std::io::Result<Vec<u8>> {
+		self.write_request(request).await?;
+		self.read_response().await
+	}
+
+	/// Writes an HTTP request message.
+	async fn write_request(&mut self, request: &str) -> std::io::Result<()> {
+		#[cfg(feature = "tokio")]
+		{
+			self.stream.write_all(request.as_bytes()).await?;
+			self.stream.flush().await
+		}
+		#[cfg(not(feature = "tokio"))]
+		{
+			self.stream.write_all(request.as_bytes())?;
+			self.stream.flush()
+		}
+	}
+
+	/// Reads an HTTP response message.
+	async fn read_response(&mut self) -> std::io::Result<Vec<u8>> {
+		#[cfg(feature = "tokio")]
+		let stream = self.stream.split().0;
+		#[cfg(not(feature = "tokio"))]
+		let stream = std::io::Read::by_ref(&mut self.stream);
+
+		let limited_stream = stream.take(MAX_HTTP_MESSAGE_HEADER_SIZE as u64);
+
+		#[cfg(feature = "tokio")]
+		let mut reader = tokio::io::BufReader::new(limited_stream);
+		#[cfg(not(feature = "tokio"))]
+		let mut reader = std::io::BufReader::new(limited_stream);
+
+		macro_rules! read_line { () => { {
+			let mut line = String::new();
+			#[cfg(feature = "tokio")]
+			let bytes_read = reader.read_line(&mut line).await?;
+			#[cfg(not(feature = "tokio"))]
+			let bytes_read = reader.read_line(&mut line)?;
+
+			match bytes_read {
+				0 => None,
+				_ => {
+					// Remove trailing CRLF
+					if line.ends_with('\n') { line.pop(); if line.ends_with('\r') { line.pop(); } }
+					Some(line)
+				},
+			}
+		} } }
+
+		// Read and parse status line
+		let status_line = read_line!()
+			.ok_or(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "no status line"))?;
+		let status = HttpStatus::parse(&status_line)?;
+
+		// Read and parse relevant headers
+		let mut message_length = HttpMessageLength::Empty;
+		loop {
+			let line = read_line!()
+				.ok_or(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "no headers"))?;
+			if line.is_empty() { break; }
+
+			let header = HttpHeader::parse(&line)?;
+			if header.has_name("Content-Length") {
+				let length = header.value.parse()
+					.map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
+				if let HttpMessageLength::Empty = message_length {
+					message_length = HttpMessageLength::ContentLength(length);
+				}
+				continue;
+			}
+
+			if header.has_name("Transfer-Encoding") {
+				message_length = HttpMessageLength::TransferEncoding(header.value.into());
+				continue;
+			}
+		}
+
+		if !status.is_ok() {
+			// TODO: Handle 3xx redirection responses.
+			return Err(std::io::Error::new(std::io::ErrorKind::NotFound, "not found"));
+		}
+
+		// Read message body
+		let read_limit = MAX_HTTP_MESSAGE_BODY_SIZE - reader.buffer().len();
+		reader.get_mut().set_limit(read_limit as u64);
```

Comment:
In the chunked encoding case, we're limiting to 4MB *including* chunked encoding overhead, which I guess isn't ideal - maybe we should just bump `MAX_HTTP_MESSAGE_BODY_SIZE` to 2*4MB + 32KB or something larger so that its clearly fine?

Reply:
Good point. Updated as recommended.